flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -1948,9 +1951,23 @@ public CalciteException handleUnresolvedFunction(
         if (unresolvedFunction instanceof SqlFunction) {
             final SqlOperandTypeChecker typeChecking =
                     new AssignableOperandTypeChecker(argTypes, argNames);
-            signature =
+            final String invocation =
                     typeChecking.getAllowedSignatures(
                             unresolvedFunction, unresolvedFunction.getName());
+            if (unresolvedFunction.getOperandTypeChecker() != null) {
+                final String allowedSignatures =
+                        unresolvedFunction
+                                .getOperandTypeChecker()
+                                .getAllowedSignatures(
+                                        unresolvedFunction, unresolvedFunction.getName());
+                throw newValidationError(
+                        call,
+                        EXTRA_RESOURCE.validatorNoFunctionMatch(invocation, allowedSignatures));
+            } else {
+                signature =
+                        typeChecking.getAllowedSignatures(
+                                unresolvedFunction, unresolvedFunction.getName());
+            }
```

Comment:
This class is copied from Apache Calcite to work around some issues mentioned in comments at the beginning of class declaration.
It is better to keep it as is to simplify future upgrades or each change (which is not done in corresponding Calcite's version should be highlighted with comments like done for others)

Reply:
+1

Reply:
Also could you please create a ticket for that improvement in Calcite's JIRA?

Reply:
Created Calcite Ticket - https://issues.apache.org/jira/browse/CALCITE-6069

Reply:
Also added comments like other modifications made. Once its implemented in Calcite, we can remove this modification.